File: _stats_build/statistics/wr_wpa.rb
require_relative "abstract/round_metric"

class WrWpa < RoundMetric
  def initialize
    @title = "WPA (Worst Possible Average)"
    @note = "Worst Possible Average: average of the worst 3 out of the first 4 solves in a round."
    @table_header = { "Result" => :right, "Improvement" => :right, "Days" => :right, "Person" => :left, "Date" => :left, "Competition" => :left, "Details" => :left }
  end

  # NOTE: 基于 5 把的指标，不适用于 Mo3 项目
  def target_events = Events::WITH_AO5

  # NOTE: WPA = 前4次中取最差的3次求均值
  # 需要全部4次都有效（> 0）
  def compute_metric(values, r)
    first4 = values[0..3]
    return nil unless first4.all? { |v| v > 0 }
    # 取最差的3个（最大值）
    worst3 = first4.sort.last(3)
    worst3.sum.to_f / 3
  end
end

File: core/packages/stats-build/src/statistics/wr_wpa.ts
// NOTE: WPA (Worst Possible Average)——前 4 次中取最差 3 次均值
// 与 Ruby _stats_build/statistics/wr_wpa.rb 1:1 对应
import { RoundMetric } from '../core/round_metric.js';
import { EVENTS_WITH_AO5 } from '../core/events.js';

export class WrWpa extends RoundMetric {
  constructor() {
    super();
    this.title = 'WPA';
    this.titleZh = 'WPA';
    this.note = 'Worst Possible Average: average of the worst 3 out of the first 4 solves in a round.';
    this.noteZh = '最差可能平均：前 4 次中取最差 3 次的平均。';
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
  }

  targetEvents() { return EVENTS_WITH_AO5; }

  // NOTE: WPA = 前 4 次中取最差 3 次均值，需要全部 4 次有效
  computeMetric(values: number[]): number | null {
    const first4 = values.slice(0, 4);
    if (!first4.every(v => v > 0)) return null;
    const worst3 = first4.sort((a, b) => b - a).slice(0, 3);
    return worst3.reduce((s, v) => s + v, 0) / 3;
  }
}
